homeassistant/components/fritzbox/config_flow.py

```diff
@@ -0,0 +1,135 @@
+DATA_SCHEMA_USER = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default=DEFAULT_HOST): str,
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+DATA_SCHEMA_CONFIRM = vol.Schema(
+    {vol.Required(CONF_USERNAME): str, vol.Required(CONF_PASSWORD): str}
+)
+
+RESULT_AUTH_MISSING = "auth_missing"
+RESULT_SUCCESS = "success"
+RESULT_NOT_FOUND = "not_found"
+
+
+def _get_ip(host):
+    if host is None:
+        return None
+    return socket.gethostbyname(host)
+
+
+class FritzboxConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a AVM Fritz!Box config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host = None
+        self._manufacturer = None
+        self._model = None
+        self._name = None
+        self._password = None
+        self._username = None
+
+    def _get_entry(self):
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                CONF_HOST: self._host,
+                CONF_PASSWORD: self._password,
+                CONF_USERNAME: self._username,
+            },
+        )
+
+    def _try_connect(self):
+        """Try to connect and check auth."""
+        fritzbox = Fritzhome(
+            host=self._host, user=self._username, password=self._password
+        )
+        try:
+            fritzbox.login()
+            fritzbox.logout()
+            return RESULT_SUCCESS
+        except OSError:
+            return RESULT_NOT_FOUND
+        except LoginError:
+            return RESULT_AUTH_MISSING
+
+    async def async_step_import(self, user_input=None):
+        """Handle configuration by yaml file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            ip_address = await self.hass.async_add_executor_job(
+                _get_ip, user_input[CONF_HOST]
+            )
+
+            await self.async_set_unique_id(ip_address)
+            self._abort_if_unique_id_configured()
+
+            self._host = user_input[CONF_HOST]
+            self._name = user_input[CONF_HOST]
+            self._password = user_input[CONF_PASSWORD]
+            self._username = user_input[CONF_USERNAME]
+
+            result = await self.hass.async_add_executor_job(self._try_connect)
+
+            if result != RESULT_SUCCESS:
+                return self.async_abort(reason=result)
+            return self._get_entry()
+
+        return self.async_show_form(step_id="user", data_schema=DATA_SCHEMA_USER)
+
+    async def async_step_ssdp(self, user_input=None):
+        """Handle a flow initialized by discovery."""
+        host = urlparse(user_input[ATTR_SSDP_LOCATION]).hostname
```

Comment:
This will error if user input is `None`.

Reply:
Can this ever happen at step ssdp?

Reply:
We can make user_input a positional argument instead.

Reply:
OK, coming from `ssdp` integration it wont be None, if I understood the code right. So I make it positional